// @powersync/react-native/lib/db/adapters/react-native-quick-sqlite/RNQSDBAdapter.js
import { BaseObserver, DBGetUtilsDefaultMixin, DBAdapterDefaultMixin } from '@powersync/common';
class RNQSConnectionPool extends BaseObserver {
    baseDB;
    name;
    constructor(baseDB, name) {
        super();
        this.baseDB = baseDB;
        this.name = name;
        // link table update commands
        baseDB.registerTablesChangedHook((update) => {
            this.iterateListeners((cb) => cb.tablesUpdated?.(update));
        });
    }
    close() {
        return this.baseDB.close();
    }
    readLock(fn, options) {
        return this.baseDB.readLock((dbTx) => fn(this.generateContext(dbTx)), options);
    }
    writeLock(fn, options) {
        return this.baseDB.writeLock((dbTx) => fn(this.generateContext(dbTx)), options);
    }
    generateContext(ctx) {
        return new QuickSqliteContext(ctx);
    }
    async refreshSchema() {
        await this.baseDB.refreshSchema();
    }
}
class QuickSqliteExecutor {
    context;
    constructor(context) {
        this.context = context;
    }
    execute(query, params) {
        return this.context.execute(query, params);
    }
    /**
     * 'executeRaw' is not implemented in RNQS, this falls back to 'execute'.
     */
    async executeRaw(query, params) {
        const result = await this.context.execute(query, params);
        const rows = result.rows?._array ?? [];
        return rows.map((row) => Object.values(row));
    }
}
class QuickSqliteContext extends DBGetUtilsDefaultMixin(QuickSqliteExecutor) {
}
/**
 * Adapter for React Native Quick SQLite
 */
export class RNQSDBAdapter extends DBAdapterDefaultMixin(RNQSConnectionPool) {
    // We don't want the default implementation here, RNQS does not support executeBatch for lock contexts so that would
    // be less efficient.
    async executeBatch(query, params = []) {
        const commands = [];
        for (let i = 0; i < params.length; i++) {
            commands.push([query, params[i]]);
        }
        const result = await this.baseDB.executeBatch(commands);
        return {
            rowsAffected: result.rowsAffected ? result.rowsAffected : 0
        };
    }
}
