// @powersync/react/lib/hooks/watched/useQuery.js
import { usePowerSync } from '../PowerSyncContext.js';
import { useSingleQuery } from './useSingleQuery.js';
import { useWatchedQuery } from './useWatchedQuery.js';
import { constructCompatibleQuery } from './watch-utils.js';
import { useAllSyncStreamsHaveSynced } from '../streams.js';
export function useQuery(query, parameters = [], options = {}) {
    const powerSync = usePowerSync();
    if (!powerSync) {
        return {
            ..._loadingState,
            isLoading: false,
            error: new Error('PowerSync not configured.')
        };
    }
    const { parsedQuery, queryChanged } = constructCompatibleQuery(query, parameters, options);
    const streamsHaveSynced = useAllSyncStreamsHaveSynced(powerSync, options?.streams);
    const runOnce = options?.runQueryOnce == true;
    const single = useSingleQuery({
        query: parsedQuery,
        powerSync,
        queryChanged,
        active: runOnce && streamsHaveSynced
    });
    const watched = useWatchedQuery({
        query: parsedQuery,
        powerSync,
        queryChanged,
        options: {
            reportFetching: options.reportFetching,
            // Maintains backwards compatibility with previous versions
            // Differentiation is opt-in by default
            // We emit new data for each table change by default.
            rowComparator: options.rowComparator
        },
        active: !runOnce && streamsHaveSynced
    });
    if (!streamsHaveSynced) {
        return { ..._loadingState };
    }
    return (runOnce ? single : watched) ?? _loadingState;
}
const _loadingState = { isLoading: true, isFetching: false, data: [], error: undefined };
